### PR TITLE
PERF: fix #32976 slow group by for categorical columns

### LINKED ISSUE #32976: Categorical columns are slow in groupby operations
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
import random

SIZE = 100000
GROUPS = 10000  # The larger, the more extreme the timing differences
CARDINALITY = 10
CAT = pd.CategoricalDtype([i for i in range(CARDINALITY)])
df_int = pd.DataFrame({
    'group': [random.randint(0, GROUPS) for i in range(SIZE)],
    'cat': [random.choice(CAT.categories) for i in range(SIZE)],
})
df_cat = df_int.astype({'cat': CAT})
df_str = df_cat.astype({'cat': 'str'})

%timeit df_int.groupby('group').last()
# 100 loops, best of 3: 5.51 ms per loop
%timeit df_str.groupby('group').last()
# 100 loops, best of 3: 11.4 ms per loop
%timeit df_cat.groupby('group').last()
# 1 loop, best of 3: 4.62 s per loop
```

#### Problem description

We have a dataframe like the above in production and noticed a particular groupby call was drastically slowing down our pipeline. We were able to workaround the issue by converting categorical types to strings before the groupby operation.

The issue occurs when categorical values are in the grouped frame (not when they're included in the groupby index). The problem scales with the number of groups. The larger the GROUPS value in the snippet above, the worse the timing differences are.

We didn't expect using categories to cause these kind of performance issues. Is this expected, or could it be something worth optimizing for?

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.6.9.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.14.137+
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 0.25.3
numpy            : 1.18.2
pytz             : 2018.9
dateutil         : 2.8.1
pip              : 19.3.1
setuptools       : 46.0.0
Cython           : 0.29.15
pytest           : 3.6.4
hypothesis       : None
sphinx           : 1.8.5
blosc            : None
feather          : 0.4.0
xlsxwriter       : None
lxml.etree       : 4.2.6
html5lib         : 1.0.1
pymysql          : None
psycopg2         : 2.7.6.1 (dt dec pq3 ext lo64)
jinja2           : 2.11.1
IPython          : 5.5.0
pandas_datareader: 0.7.4
bs4              : 4.6.3
bottleneck       : 1.3.2
fastparquet      : None
gcsfs            : None
lxml.etree       : 4.2.6
matplotlib       : 3.2.1
numexpr          : 2.7.1
odfpy            : None
openpyxl         : 2.5.9
pandas_gbq       : 0.11.0
pyarrow          : 0.14.1
pytables         : None
s3fs             : 0.4.0
scipy            : 1.4.1
sqlalchemy       : 1.3.15
tables           : 3.4.4
xarray           : 0.15.0
xlrd             : 1.1.0
xlwt             : 1.3.0
xlsxwriter       : None
</details>

### PR DESCRIPTION
**Pull request by rtlee9**
Aggregate categorical codes with fast cython aggregation for select `how` operations. Added new ASV benchmark copied from 32976 indicating > 99% improvement in performance for this case.

- [x] closes #32976
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/groupby/generic.py [hunk 1]
```diff
@@ -1056,9 +1058,47 @@ def _cython_agg_blocks(
             result = no_result
             locs = block.mgr_locs.as_array
             try:
-                result, _ = self.grouper.aggregate(
-                    block.values, how, axis=1, min_count=min_count
+
+                cat_method_blacklist = (
```
Comment (rtlee9) [inline pandas/core/groupby/generic.py:1062]: I copied this list from `asv_bench.benchmarks.groupby.method_blacklist['object']` and appended the "add" method. Is there a better way of blacklisting these methods which shouldn't be applied to categorical codes?

`tests.groupby.test_function.test_arg_passthru` is an example of a test failure without this blacklisting.

File: pandas/core/groupby/generic.py [hunk 2]
```diff
@@ -1056,9 +1058,47 @@ def _cython_agg_blocks(
...
+                    "cummax",
+                    "rank",
+                    "pct_change",
+                    "min",
+                    "var",
+                    "mad",
+                    "describe",
+                    "std",
+                    "quantile",
                 )
```
Comment (jbrockmendel) [inline pandas/core/groupby/generic.py:1083]: pls define this before the "try"

File: pandas/core/groupby/generic.py [hunk 3]
```diff
@@ -1056,9 +1058,47 @@ def _cython_agg_blocks(
...
+                    # perf improvement: aggregate categorical codes with fast cython agg
+                    # for select `how` operations
+                    result_codes, _ = self.grouper.aggregate(
+                        block.values.codes, how, axis=1, min_count=min_count
+                    )
+                    result = Categorical.from_codes(
+                        result_codes,
+                        categories=block.values.categories,
+                        ordered=block.values.ordered,
+                    )
```
Comment (TomAugspurger) [inline pandas/core/groupby/generic.py:1097]:
```suggestion
                        dtype=block.values.dtype
                    )
```

File: doc/source/whatsnew/v1.1.0.rst [hunk 4]
```diff
@@ -426,6 +426,7 @@ Performance improvements
   :meth:`DataFrame.sparse.from_spmatrix` constructor (:issue:`32821`,
   :issue:`32825`,  :issue:`32826`, :issue:`32856`, :issue:`32858`).
 - Performance improvement in reductions (sum, prod, min, max) for nullable (integer and boolean) dtypes (:issue:`30982`, :issue:`33261`, :issue:`33442`).
+- Performance improvement in :meth:`DataFrameGroupBy.func` when aggregating categorical data (:issue:`32976`)
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v1.1.0.rst:429]:
```suggestion
- Performance improvement in :meth:`DataFrame.groupby` when aggregating categorical data (:issue:`32976`)
```

**Review by TomAugspurger — COMMENTED** [review]
@jbrockmendel how close do you think we are to defining an API to allow passing EA values to cython arrays?


```python
# ignore the method names.
values = self._values_for_cython(method="first")
# ... operation on values
result = self._result_from_cython(..., dtype=self.dtype method="first")
```

For categorical, that method returns `codes` and `result_from_cython` is `from_codes`.

**Comment by jbrockmendel** [comment]
> how close do you think we are to defining an API to allow passing EA values to cython arrays?
> `values = self._values_for_cython(method="first")`

If we restrict attention to ordering-based cython methods then i think this is pretty straightforward, just need to get everyone on board.

File: asv_bench/benchmarks/groupby.py [hunk 5]
```diff
@@ -510,6 +511,27 @@ def time_groupby_extra_cat_nosort(self):
         self.df_extra_cat.groupby("a", sort=False)["b"].count()
 
 
+class CategoricalFrame:
+    def setup(self):
```
Comment (jreback) [inline asv_bench/benchmarks/groupby.py:515]: how is this different than the section right above?

File: pandas/core/groupby/generic.py [hunk 6]
```diff
@@ -1056,9 +1058,47 @@ def _cython_agg_blocks(
             result = no_result
             locs = block.mgr_locs.as_array
             try:
-                result, _ = self.grouper.aggregate(
-                    block.values, how, axis=1, min_count=min_count
+
```
Comment (jreback) [inline pandas/core/groupby/generic.py:1061]: please don't do this here. this patch needs to be much lower in the stack. follow .aggregate thru and see where other types of groupers are handled.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/generic.py (hunk 3 above)
Comment (rtlee9) [reply pandas/core/groupby/generic.py:1097]: I actually tried this and found that it raises a ValueError [here](https://github.com/pandas-dev/pandas/blob/1c0cc62e3/pandas/core/dtypes/dtypes.py#L350): Cannot specify `categories` or `ordered` together with `dtype`

File: asv_bench/benchmarks/groupby.py (hunk 5 above)
Comment (rtlee9) [reply asv_bench/benchmarks/groupby.py:515]: the values are categories in this new test

File: pandas/core/groupby/generic.py (hunk 6 above)
Comment (rtlee9) [reply pandas/core/groupby/generic.py]: this is two calls further in the stack (where the NotImplementedError is raised to start with). look like a better place for this patch?
https://github.com/pandas-dev/pandas/blob/1c0cc62e3/pandas/core/groupby/ops.py#L455

Comment (rtlee9) [reply pandas/core/groupby/generic.py]: Moved in commit `0340554e2`

File: asv_bench/benchmarks/groupby.py (hunk 5 above)
Comment (jreback) [reply asv_bench/benchmarks/groupby.py:515]: can you add a comment to this class to that effect (and to the above that its just for grouping.

File: doc/source/whatsnew/v1.1.0.rst (hunk 4 above)
Comment (jreback) [reply doc/source/whatsnew/v1.1.0.rst:430]: can you update as @TomAugspurger indicated (and don't just resolve)

File: pandas/core/groupby/ops.py [hunk 7]
```diff
@@ -472,6 +473,29 @@ def _cython_operation(
 
         is_datetimelike = needs_i8_conversion(values.dtype)
         is_numeric = is_numeric_dtype(values.dtype)
+        is_categorical = is_categorical_dtype(values)
+        cat_method_blacklist = (
+            "add",
```
Comment (jreback) [inline pandas/core/groupby/ops.py:478]: why do we need a blacklist at all? you are already only operating on the codes

File: pandas/tests/groupby/aggregate/test_aggregate.py [hunk 8]
```diff
@@ -466,7 +466,7 @@ def test_agg_cython_category_not_implemented_fallback():
     result = df.groupby("col_num").col_cat.first()
     expected = pd.Series(
         [1, 2, 3], index=pd.Index([1, 2, 3], name="col_num"), name="col_cat"
-    )
+    ).astype("category")
```
Comment (jreback) [inline pandas/tests/groupby/aggregate/test_aggregate.py:469]: why is this changed?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
pls show the performance before / after.

File: asv_bench/benchmarks/groupby.py [hunk 9]
```diff
@@ -510,6 +511,27 @@ def time_groupby_extra_cat_nosort(self):
...
+        CAT = CategoricalDtype(list(range(CARDINALITY)))
+        df_int = DataFrame(
+            {
+                "group": [np.random.randint(0, GROUPS) for i in range(SIZE)],
+                "cat": [np.random.choice(CAT.categories) for i in range(SIZE)],
+            }
+        )
+        self.df_cat_values = df_int.astype({"cat": CAT})
+
+    def time_groupby(self):
```
Comment (jreback) [inline asv_bench/benchmarks/groupby.py:528]: likely need to paramaterize this over a bunch of functions. assume this is not too slow to do that. if so pls reduce the size of the benchmark to make it reasonable to parameterize.

**Comment by rtlee9** [comment]
latest asv results for new benchmark replicating #32976:

```
$ asv continuous HEAD -b 'groupby.CategoricalFrame'

before           after         ratio    
     [77a0f19c]       [ec70c57e]
     <master>         <cat_groupby_fix>
-      1.68±0.01s       4.51±0.1ms     0.00  groupby.CategoricalFrame.time_groupby
-      1.68±0.01s      4.43±0.09ms     0.00  groupby.CategoricalFrame.time_groupby_ordered
                                                                 
SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.          
PERFORMANCE INCREASED.
```

File: asv_bench/benchmarks/groupby.py (hunk 9 above)
Comment (rtlee9) [reply asv_bench/benchmarks/groupby.py:530]: this takes 10.2 ms ± 28.7 µs per loop on the current commit on my desktop via timeit but 7.14 s ± 31.3 ms per loop on master.  I think 10ms should be ok but please let me know otherwise

File: pandas/core/groupby/ops.py (hunk 7 above)
Comment (rtlee9) [reply pandas/core/groupby/ops.py:478]: If this method is passed categorical values with a "how" like mean then we don't want to average the codes as if they were ints -- this would happen in the `pandas.tests.groupby.test_function.test_arg_passthru` test without the blacklist, for example. the logic for cases these blacklist cases is already handled in higher in the call stack in `pandas.core.groupby.generic.DataFrameGroupBy._cython_agg_blocks` but maybe we could move some of that logic here for conciseness and/or perf later on

File: doc/source/whatsnew/v1.1.0.rst (hunk 4 above)
Comment (rtlee9) [reply doc/source/whatsnew/v1.1.0.rst]: my bad, meant to amend that last time around

File: asv_bench/benchmarks/groupby.py (hunk 9 above)
Comment (jreback) [reply asv_bench/benchmarks/groupby.py:530]: yes this is fine, but i need you to paramaterize over multiple functions, doesn't have to be every one but represententative ones (e.g. reductions, transforms and filters)

File: pandas/core/groupby/ops.py (hunk 7 above)
Comment (jreback) [reply pandas/core/groupby/ops.py:478]: i don't think you need this at all, what breaks if you take it out entirely (the blackist)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/ops.py (hunk 7 above)
Comment (rtlee9) [reply pandas/core/groupby/ops.py:478]: This test will fail without the blacklist: `pandas.tests.groupby.test_function.test_arg_passthru`. Here's a minimal example that will fail:

```python
import pandas as pd

df = pd.DataFrame(
    {"group": [1, 1, 2], "category_string": pd.Series(list("abc")).astype("category")},
    columns=["group", "category_string"],
)

df.groupby("group").mean(numeric_only=False)
```

Nothing else fails so perhaps this test can be updated in lieu of the blacklist, not sure if this is fine from an interface perspective.

File: asv_bench/benchmarks/groupby.py (hunk 9 above)
Comment (rtlee9) [reply asv_bench/benchmarks/groupby.py]: paramaterized over column types in b4648d5a0

File: pandas/tests/groupby/aggregate/test_aggregate.py (hunk 8 above)
Comment (rtlee9) [reply pandas/tests/groupby/aggregate/test_aggregate.py:469]: The `_cython_operation` method in this commit returns a categorical if passed a categorical. In this test the "col_cat" column is categorical so it seems intuitive to me that when aggregated over it should also be a category.

The fact that this test was passing before suggests that this is not also true in the higher order functions like `DataFrameGroupBy._cython_agg_blocks`.

File: asv_bench/benchmarks/groupby.py [hunk 10]
```diff
@@ -510,6 +512,33 @@ def time_groupby_extra_cat_nosort(self):
         self.df_extra_cat.groupby("a", sort=False)["b"].count()
 
 
+class CategoricalFrame:
+    # benchmark grouping with operations on categorical values (GH #32976)
+    param_names = ["groupby_type", "value_type"]
```
Comment (jreback) [inline asv_bench/benchmarks/groupby.py:517]: oarameterize over 

mean and head as well

i would also reduce the number of groups as well 
it will still show an appreciable diff

Comment (rtlee9) [reply asv_bench/benchmarks/groupby.py]: done. but I swapped mean for count since mean requires numeric types

Comment (rtlee9) [reply asv_bench/benchmarks/groupby.py]: i also removed the str groupby parameter option to limit the number of tests but the benchmark results are the same either way for this diff.

here are the latest results:
```
       before           after         ratio
     [22cf0f5d]       [a16f6a23]
     <master>                   
-         167±1ms      2.72±0.09ms     0.02  groupby.CategoricalFrame.time_groupby(<class 'int'>, <class 'str'>, 'last')
-         168±2ms      2.66±0.07ms     0.02  groupby.CategoricalFrame.time_groupby_ordered(<class 'int'>, <class 'str'>, 'last')
-         174±2ms      2.65±0.09ms     0.02  groupby.CategoricalFrame.time_groupby_ordered(<class 'int'>, <class 'int'>, 'last')
-         175±2ms      2.61±0.05ms     0.01  groupby.CategoricalFrame.time_groupby(<class 'int'>, <class 'int'>, 'last')

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

File: pandas/tests/groupby/aggregate/test_aggregate.py (hunk 8 above)
Comment (TomAugspurger) [reply pandas/tests/groupby/aggregate/test_aggregate.py:469]: This change needs a dedicated whatsnew entry.

Agreed that preserving the dtype is probably the right choice, but this is a relatively large change.

Comment (rtlee9) [reply pandas/tests/groupby/aggregate/test_aggregate.py:469]: Sure, just added a new entry in the "Other API changes" section in commit 6498d6ba2

File: pandas/core/groupby/ops.py (hunk 7 above)
Comment (jreback) [reply pandas/core/groupby/ops.py:488]: yeah i think its ok to simply update this test

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/ops.py (hunk 7 above)
Comment (TomAugspurger) [reply pandas/core/groupby/ops.py:488]: It really seems like we need this blacklist anytime the cython op does something with the actual *values* then we can't pass the codes there. For example something like `mean` or `sum` on a numeric categorical.

Comment (jreback) [reply pandas/core/groupby/ops.py:488]: I see. ok then, let's move this to an attribute on the class, or a function maybe better for easier re-sue

Comment (rtlee9) [reply pandas/core/groupby/ops.py]: Moved to class level in 79f0c72. We could easily make this accessible outside the class with an additional method w/ property decorator

**Comment by TomAugspurger** [comment]
@rtlee9 Can you merge master and fix the conflict?

**Comment by rtlee9** [comment]
Fixed conflict and merged in 7570425

File: doc/source/whatsnew/v1.1.0.rst [hunk 11]
```diff
@@ -239,7 +239,7 @@ Other API changes
 - ``loc`` lookups with an object-dtype :class:`Index` and an integer key will now raise ``KeyError`` instead of ``TypeError`` when key is missing (:issue:`31905`)
 - Using a :func:`pandas.api.indexers.BaseIndexer` with ``count``, ``min``, ``max``, ``median``, ``skew``,  ``cov``, ``corr`` will now return correct results for any monotonic :func:`pandas.api.indexers.BaseIndexer` descendant (:issue:`32865`)
 - Added a :func:`pandas.api.indexers.FixedForwardWindowIndexer` class to support forward-looking windows during ``rolling`` operations.
--
+- :meth:`DataFrame.groupby` aggregations of categorical series will now return a :class:`Categorical` while preserving the codes and categories of the original series
```
Comment (jreback) [inline doc/source/whatsnew/v1.1.0.rst:242]: is this some other issue?

File: pandas/core/groupby/ops.py [hunk 12]
```diff
@@ -494,6 +519,17 @@ def _cython_operation(
                 values = ensure_int_or_float(values)
         elif is_numeric and not is_complex_dtype(values):
             values = ensure_float64(values)
+        elif is_categorical:
+            if how in self._cat_method_blacklist:
```
Comment (jreback) [inline pandas/core/groupby/ops.py:523]: I really don't like doing this. Can you elaborate when we can actually process this? listing methods is a bad idea generally.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
@rtlee9 if you'd address the comments and merge master

File: doc/source/whatsnew/v1.1.0.rst (hunk 11 above)
Comment (rtlee9) [reply doc/source/whatsnew/v1.1.0.rst:242]: This is a (desirable) side effect of this PR - please find more context in [this](https://github.com/pandas-dev/pandas/pull/33739#discussion_r416344666) thread.

File: pandas/core/groupby/ops.py (hunk 12 above)
Comment (rtlee9) [reply pandas/core/groupby/ops.py:523]: I'm not exactly sure what you mean by "when we can actually process this" this but I agree that listing methods isn't necessarily thorough and isn't robust. However, I've been unable to find a suitable alternative to blacklisting methods where we don't want to apply the aggregation on the category codes -- open to other ideas though. Please find more context in [this](https://github.com/pandas-dev/pandas/pull/33739#discussion_r415151884) thread.

**Review by jreback — CHANGES_REQUESTED** [review]
if u can merge master

i don't really like the specific method blacklist - if u can find a better way

**Comment by simonjayhawkins** [comment]
@rtlee9 can you move release note to 1.2 and merge upstream/master to resolve conflict

File: doc/source/whatsnew/v1.2.0.rst [hunk 13]
```diff
@@ -132,6 +133,7 @@ Plotting
 Groupby/resample/rolling
 ^^^^^^^^^^^^^^^^^^^^^^^^
 
+- :meth:`DataFrame.groupby` aggregations of categorical series will now return a :class:`Categorical` while preserving the codes and categories of the original series
```
Comment (simonjayhawkins) [inline doc/source/whatsnew/v1.2.0.rst:136]: can you add this PR as the issue (or relevant issue if discussed elsewhere.

Comment (rtlee9) [reply doc/source/whatsnew/v1.2.0.rst]: Added the PR as the issue in commit 0fd3dde3f

**Comment by dsaxton** [comment]
@rtlee9 Is this still active? Some merge conflicts that need resolving.

File: pandas/core/groupby/ops.py [hunk 14]
```diff
@@ -356,6 +357,29 @@ def get_group_levels(self) -> List[Index]:
 
     _name_functions = {"ohlc": ["open", "high", "low", "close"]}
 
+    _cat_method_blacklist = (
+        "add",
```
Comment (jreback) [inline pandas/core/groupby/ops.py:361]: what methods *don't* work?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by arw2019** [comment]
Closing for now. @rtlee9 ping us whenever you'd like to continue and we'll reopen!
